Extension/browser/chrome/lib/utils/local-storage.js

```diff
@@ -15,32 +15,149 @@
+                    var result = null;
+                    if (results) {
+                        result = results[path];
+                    }
+                    callback(null, result);
+                }
+            });
+        } catch (ex) {
+            callback(ex);
+        }
+    }
+
+    /**
+     * Writes data to storage.local
+     * @param path Path
+     * @param data Data to write
+     * @param callback Callback
+     */
+    function write(path, data, callback) {
+        var item = {};
+        item[path] = data;
+        try {
+            browser.storage.local.set(item, function () {
+                if (!checkLastError(callback)) {
+                    callback();
+                }
+            });
+        } catch (ex) {
+            callback(ex);
+        }
+    }
+
+    /**
+     * Migrates key-value pair from local storage to storage.local
+     * Part of task https://github.com/AdguardTeam/AdguardBrowserExtension/issues/681
+     * @param key Key to migrate
+     */
+    function migrateKeyValue(key) {
+        if (key in localStorage) {
+            var value = localStorage.getItem(key);
+            localStorage.removeItem(key);
+            setItem(key, value);
+        }
+    }
+
+    /**
+     * Retrieves value by key from cached values
+     * @param key
+     * @returns {*}
+     */
     var getItem = function (key) {
-        return localStorage.getItem(key);
+        if (!(key in values)) {
+            migrateKeyValue(key);
+        }
+        return values[key];
     };
 
     var setItem = function (key, value) {
-        localStorage.setItem(key, value);
+        values[key] = value;
+        write(ADGUARD_SETTINGS_PROP, values, checkError);
     };
 
     var removeItem = function (key) {
+        delete values[key];
+        // Remove from localStorage too, as a part of migration process
         localStorage.removeItem(key);
+        write(ADGUARD_SETTINGS_PROP, values, checkError);
     };
 
     var hasItem = function (key) {
-        return key in localStorage;
+        if (key in values) {
+            return true;
+        }
+        migrateKeyValue(key);
+        return key in values;
+    };
+
+    /**
+     * We can't use localStorage object anymore and we've decided to store all data into storage.local
+     * localStorage is affected by cleaning tools: https://github.com/AdguardTeam/AdguardBrowserExtension/issues/681
+     * storage.local has async nature and we have to preload all key-values pairs into memory on extension startup
+     *
+     * @param callback
+     */
+    var init = function (callback) {
+        if (values !== null) {
+            // Already initialized
+            callback();
+            return;
+        }
+        read(ADGUARD_SETTINGS_PROP, function (ex, items) {
+            if (ex) {
+                checkError(ex);
+            }
+            values = items || Object.create(null);
+            callback();
+        });
```

Comment:
The only thing we can do if settings read failed is to write this error to log and startup extension as new. 
Any other proposals?